File: storm-java21/src/main/java/st/orm/repository/impl/ProjectionRepositoryImpl.java
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.repository.impl;

import static java.util.Objects.requireNonNull;
import static st.orm.template.impl.StringTemplates.convert;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;
import st.orm.Data;
import st.orm.Metamodel;
import st.orm.Page;
import st.orm.Pageable;
import st.orm.Projection;
import st.orm.Ref;
import st.orm.repository.ProjectionRepository;
import st.orm.template.Model;
import st.orm.template.ORMTemplate;
import st.orm.template.QueryBuilder;
import st.orm.template.impl.ModelImpl;
import st.orm.template.impl.ORMTemplateImpl;
import st.orm.template.impl.QueryBuilderImpl;

/**
 */
public final class ProjectionRepositoryImpl<P extends Projection<ID>, ID> implements ProjectionRepository<P, ID> {
    private final st.orm.core.repository.ProjectionRepository<P, ID> core;

    public ProjectionRepositoryImpl(st.orm.core.repository.ProjectionRepository<P, ID> core) {
        this.core = requireNonNull(core);
    }

    @Override
    public Model<P, ID> model() {
        return new ModelImpl<>((st.orm.core.template.impl.ModelImpl<P, ID>) core.model());
    }

    @Override
    public Ref<P> ref(ID id) {
        return core.ref(id);
    }

    @Override
    public Ref<P> ref(P projection, ID id) {
        return core.ref(projection, id);
    }

    @Override
    public QueryBuilder<P, P, ID> select() {
        return new QueryBuilderImpl<>(core.select());
    }

    @Override
    public QueryBuilder<P, Long, ID> selectCount() {
        return new QueryBuilderImpl<>(core.selectCount());
    }

    @Override
    public <R> QueryBuilder<P, R, ID> select(Class<R> selectType) {
        return new QueryBuilderImpl<>(core.select(selectType));
    }

    @Override
    public QueryBuilder<P, Ref<P>, ID> selectRef() {
        return new QueryBuilderImpl<>(core.selectRef());
    }

    @Override
    public <R> QueryBuilder<P, R, ID> select(Class<R> selectType, StringTemplate template) {
        return new QueryBuilderImpl<>(core.select(selectType, convert(template)));
    }

    @Override
    public <R extends Data> QueryBuilder<P, Ref<R>, ID> selectRef(Class<R> refType) {
        return new QueryBuilderImpl<>(core.selectRef(refType));
    }

    @Override
    public long count() {
        return core.count();
    }

    @Override
    public boolean exists() {
        return core.exists();
    }

    @Override
    public boolean existsById(ID id) {
        return core.existsById(id);
    }

    @Override
    public boolean existsByRef(Ref<P> ref) {
        return core.existsByRef(ref);
    }

    // Page methods.

    @Override
    public Page<P> page(int pageNumber, int pageSize) {
        return core.page(pageNumber, pageSize);
    }

    @Override
    public Page<P> page(Pageable pageable) {
        return core.page(pageable);
    }

    @Override
    public Page<Ref<P>> pageRef(int pageNumber, int pageSize) {
        return core.pageRef(pageNumber, pageSize);
    }

    @Override
    public Page<Ref<P>> pageRef(Pageable pageable) {
        return core.pageRef(pageable);
    }

    @Override
    public Optional<P> findById(ID id) {
        return core.findById(id);
    }

    @Override
    public Optional<P> findByRef(Ref<P> ref) {
        return core.findByRef(ref);
    }

    @Override
    public P getById(ID id) {
        return core.getById(id);
    }

    @Override
    public P getByRef(Ref<P> ref) {
        return core.getByRef(ref);
    }

    @Override
    public <V> Optional<P> findBy(Metamodel.Key<P, V> key, V value) {
        return core.findBy(key, value);
    }

    @Override
    public <V> P getBy(Metamodel.Key<P, V> key, V value) {
        return core.getBy(key, value);
    }

    @Override
    public <V extends Data> Optional<P> findByRef(Metamodel.Key<P, V> key, Ref<V> value) {
        return core.findByRef(key, value);
    }

    @Override
    public <V extends Data> P getByRef(Metamodel.Key<P, V> key, Ref<V> value) {
        return core.getByRef(key, value);
    }

    @Override
    public List<P> findAll() {
        return core.findAll();
    }

    @Override
    public List<Ref<P>> findAllRef() {
        return core.findAllRef();
    }

    @Override
    public List<P> findAllById(Iterable<ID> ids) {
        return core.findAllById(ids);
    }

    @Override
    public List<P> findAllByRef(Iterable<Ref<P>> refs) {
        return core.findAllByRef(refs);
    }

    @Override
    public Stream<P> selectById(Stream<ID> ids) {
        return core.selectById(ids);
    }

    @Override
    public Stream<P> selectByRef(Stream<Ref<P>> refs) {
        return core.selectByRef(refs);
    }

    @Override
    public Stream<P> selectById(Stream<ID> ids, int batchSize) {
        return core.selectById(ids, batchSize);
    }

    @Override
    public Stream<P> selectByRef(Stream<Ref<P>> refs, int batchSize) {
        return core.selectByRef(refs, batchSize);
    }

    @Override
    public long countById(Stream<ID> ids) {
        return core.countById(ids);
    }

    @Override
    public long countById(Stream<ID> ids, int batchSize) {
        return core.countById(ids, batchSize);
    }

    @Override
    public long countByRef(Stream<Ref<P>> refs) {
        return core.countByRef(refs);
    }

    @Override
    public long countByRef(Stream<Ref<P>> refs, int batchSize) {
        return core.countByRef(refs, batchSize);
    }

    @Override
    public ORMTemplate orm() {
        return new ORMTemplateImpl(core.orm());
    }
}

File: storm-kotlin/src/main/kotlin/st/orm/repository/impl/ProjectionRepositoryImpl.kt
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.repository.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.fold
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.stream.consumeAsFlow
import st.orm.Data
import st.orm.Metamodel
import st.orm.Page
import st.orm.Pageable
import st.orm.Projection
import st.orm.Ref
import st.orm.Slice
import st.orm.repository.ProjectionRepository
import st.orm.template.*
import st.orm.template.impl.ModelImpl
import st.orm.template.impl.ORMTemplateImpl
import st.orm.template.impl.QueryBuilderImpl
import kotlin.reflect.KClass

/**
 */
class ProjectionRepositoryImpl<P, ID : Any>(
    private val core: st.orm.core.repository.ProjectionRepository<P, ID>,
) : ProjectionRepository<P, ID> where P : Data, P : Projection<ID> {

    override val orm: ORMTemplate
        get() = ORMTemplateImpl(core.orm())

    override val model: Model<P, ID>
        get() = ModelImpl<P, ID>(core.model())

    override fun ref(id: ID): Ref<P> = core.ref(id)

    override fun ref(projection: P, id: ID): Ref<P> = core.ref(projection, id)

    override fun select(): QueryBuilder<P, P, ID> = QueryBuilderImpl(core.select())

    override fun selectCount(): QueryBuilder<P, Long, ID> = QueryBuilderImpl(core.selectCount())

    override fun <R : Any> select(selectType: KClass<R>): QueryBuilder<P, R, ID> = QueryBuilderImpl(core.select(selectType.java))

    override fun selectRef(): QueryBuilder<P, Ref<P>, ID> = QueryBuilderImpl<P, Ref<P>, ID>(core.selectRef())

    override fun <R : Any> select(
        selectType: KClass<R>,
        template: TemplateString,
    ): QueryBuilder<P, R, ID> = QueryBuilderImpl<P, R, ID>(
        core.select(selectType.java, template.unwrap),
    )

    override fun <R : Data> selectRef(refType: KClass<R>): QueryBuilder<P, Ref<R>, ID> = QueryBuilderImpl(core.selectRef(refType.java))

    override fun count(): Long = core.count()

    override fun exists(): Boolean = core.exists()

    override fun existsById(id: ID): Boolean = core.existsById(id)

    override fun existsByRef(ref: Ref<P>): Boolean = core.existsByRef(ref)

    override fun findById(id: ID): P? = core.findById(id).orElse(null)

    override fun findByRef(ref: Ref<P>): P? = core.findByRef(ref).orElse(null)

    override fun getById(id: ID): P = core.getById(id)

    override fun getByRef(ref: Ref<P>): P = core.getByRef(ref)

    override fun <V : Any> findBy(key: Metamodel.Key<P, V>, value: V): P? = core.findBy(key, value).orElse(null)

    override fun <V : Any> getBy(key: Metamodel.Key<P, V>, value: V): P = core.getBy(key, value)

    override fun <V : Data> findByRef(key: Metamodel.Key<P, V>, value: Ref<V>): P? = core.findByRef(key, value).orElse(null)

    override fun <V : Data> getByRef(key: Metamodel.Key<P, V>, value: Ref<V>): P = core.getByRef(key, value)

    override fun findAll(): List<P> = core.findAll()

    override fun findAllById(ids: Iterable<ID>): List<P> = core.findAllById(ids)

    override fun findAllByRef(refs: Iterable<Ref<P>>): List<P> = core.findAllByRef(refs)

    override fun selectAll(): Flow<P> = core.selectAll().consumeAsFlow()

    override fun selectById(ids: Flow<ID>): Flow<P> = ids.chunked(core.defaultChunkSize)
        .flatMapConcat { core.findAllById(it).asFlow() }

    override fun selectByRef(refs: Flow<Ref<P>>): Flow<P> = refs.chunked(core.defaultChunkSize)
        .flatMapConcat { core.findAllByRef(it).asFlow() }

    override fun selectById(ids: Flow<ID>, chunkSize: Int): Flow<P> = ids.chunked(chunkSize)
        .flatMapConcat { core.findAllById(it).asFlow() }

    override fun selectByRef(refs: Flow<Ref<P>>, chunkSize: Int): Flow<P> = refs.chunked(chunkSize)
        .flatMapConcat { core.findAllByRef(it).asFlow() }

    override suspend fun countById(ids: Flow<ID>): Long = ids.chunked(core.defaultChunkSize)
        .map { chunk -> core.countById(chunk.stream()) }
        .fold(0L) { acc, v -> acc + v }

    override suspend fun countById(ids: Flow<ID>, chunkSize: Int): Long = ids.chunked(chunkSize)
        .map { chunk -> core.countById(chunk.stream()) }
        .fold(0L) { acc, v -> acc + v }

    override suspend fun countByRef(refs: Flow<Ref<P>>): Long = refs.chunked(core.defaultChunkSize)
        .map { chunk -> core.countByRef(chunk.stream()) }
        .fold(0L) { acc, v -> acc + v }

    override suspend fun countByRef(refs: Flow<Ref<P>>, chunkSize: Int): Long = refs.chunked(chunkSize)
        .map { chunk -> core.countByRef(chunk.stream()) }
        .fold(0L) { acc, v -> acc + v }

    // Page methods.

    override fun page(pageNumber: Int, pageSize: Int): Page<P> = core.page(pageNumber, pageSize)

    override fun page(pageable: Pageable): Page<P> = core.page(pageable)

    override fun pageRef(pageNumber: Int, pageSize: Int): Page<Ref<P>> = core.pageRef(pageNumber, pageSize)

    override fun pageRef(pageable: Pageable): Page<Ref<P>> = core.pageRef(pageable)

    // Slice methods.

    override fun <V> slice(key: Metamodel.Key<P, V>, size: Int): Slice<P> = core.slice(key, size)

    override fun <V> sliceBefore(key: Metamodel.Key<P, V>, size: Int): Slice<P> = core.sliceBefore(key, size)

    override fun <V> sliceRef(key: Metamodel.Key<P, V>, size: Int): Slice<Ref<P>> = core.sliceRef(key, size)

    override fun <V> sliceBeforeRef(key: Metamodel.Key<P, V>, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, size)

    override fun <V> sliceAfter(key: Metamodel.Key<P, V>, after: V, size: Int): Slice<P> = core.sliceAfter(key, after, size)

    override fun <V> sliceAfterRef(key: Metamodel.Key<P, V>, after: V, size: Int): Slice<Ref<P>> = core.sliceAfterRef(key, after, size)

    override fun <V> sliceBefore(key: Metamodel.Key<P, V>, before: V, size: Int): Slice<P> = core.sliceBefore(key, before, size)

    override fun <V> sliceBeforeRef(key: Metamodel.Key<P, V>, before: V, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, before, size)

    override fun <V : Data> sliceAfter(key: Metamodel.Key<P, V>, after: Ref<V>, size: Int): Slice<P> = core.sliceAfter(key, after, size)

    override fun <V : Data> sliceAfterRef(key: Metamodel.Key<P, V>, after: Ref<V>, size: Int): Slice<Ref<P>> = core.sliceAfterRef(key, after, size)

    override fun <V : Data> sliceBefore(key: Metamodel.Key<P, V>, before: Ref<V>, size: Int): Slice<P> = core.sliceBefore(key, before, size)

    override fun <V : Data> sliceBeforeRef(key: Metamodel.Key<P, V>, before: Ref<V>, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, before, size)

    // Composite keyset slice methods.

    override fun <V, S> slice(key: Metamodel.Key<P, V>, sort: Metamodel<P, S>, size: Int): Slice<P> = core.slice(key, sort, size)

    override fun <V, S> sliceBefore(key: Metamodel.Key<P, V>, sort: Metamodel<P, S>, size: Int): Slice<P> = core.sliceBefore(key, sort, size)

    override fun <V, S> sliceAfter(key: Metamodel.Key<P, V>, keyAfter: V, sort: Metamodel<P, S>, sortAfter: S, size: Int): Slice<P> = core.sliceAfter(key, keyAfter, sort, sortAfter, size)

    override fun <V, S> sliceBefore(key: Metamodel.Key<P, V>, keyBefore: V, sort: Metamodel<P, S>, sortBefore: S, size: Int): Slice<P> = core.sliceBefore(key, keyBefore, sort, sortBefore, size)

    override fun <V : Data, S> sliceAfter(key: Metamodel.Key<P, V>, keyAfter: Ref<V>, sort: Metamodel<P, S>, sortAfter: S, size: Int): Slice<P> = core.sliceAfter(key, keyAfter, sort, sortAfter, size)

    override fun <V : Data, S> sliceBefore(key: Metamodel.Key<P, V>, keyBefore: Ref<V>, sort: Metamodel<P, S>, sortBefore: S, size: Int): Slice<P> = core.sliceBefore(key, keyBefore, sort, sortBefore, size)

    override fun <V, S> sliceRef(key: Metamodel.Key<P, V>, sort: Metamodel<P, S>, size: Int): Slice<Ref<P>> = core.sliceRef(key, sort, size)

    override fun <V, S> sliceBeforeRef(key: Metamodel.Key<P, V>, sort: Metamodel<P, S>, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, sort, size)

    override fun <V, S> sliceAfterRef(key: Metamodel.Key<P, V>, keyAfter: V, sort: Metamodel<P, S>, sortAfter: S, size: Int): Slice<Ref<P>> = core.sliceAfterRef(key, keyAfter, sort, sortAfter, size)

    override fun <V, S> sliceBeforeRef(key: Metamodel.Key<P, V>, keyBefore: V, sort: Metamodel<P, S>, sortBefore: S, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, keyBefore, sort, sortBefore, size)

    override fun <V : Data, S> sliceAfterRef(key: Metamodel.Key<P, V>, keyAfter: Ref<V>, sort: Metamodel<P, S>, sortAfter: S, size: Int): Slice<Ref<P>> = core.sliceAfterRef(key, keyAfter, sort, sortAfter, size)

    override fun <V : Data, S> sliceBeforeRef(key: Metamodel.Key<P, V>, keyBefore: Ref<V>, sort: Metamodel<P, S>, sortBefore: S, size: Int): Slice<Ref<P>> = core.sliceBeforeRef(key, keyBefore, sort, sortBefore, size)
}
